standard/types.md

```diff
@@ -705,3 +714,150 @@ An *unmanaged_type* is any type that isn’t a *reference_type*, a *type_paramet
+  > *Note*: This message is characterized as “informational” rather than “warning,” so as not to confuse it with the state of the nullable warning context, which is unrelated. *end note*
+- The null-forgiving operator `!` is ignored.
+
+> *Example*:
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext1", ignoredWarnings:["CS0219","CS8632"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable disable
+> string? s1 = null;    // Informational message; ? is ignored; s2 already nullable
+> string s2 = null;     // OK; null initialization of a nullable reference
+> s2 = null;            // OK; null assignment to a nullable reference
+> char c1 = s2[1];      // OK; no warning on dereference of a possible null; throws NullReferenceException
+> c1 = s2![1];          // OK; ! is ignored
+> ```
+>
+> *end example*
+
+When the nullable annotation context is enabled
+
+- For any reference type `T`, the annotation `?` in `T?` makes `T?` a nullable type, whereas the unannotated `T` is non-nullable.
+- A warning shall be generated when a variable of an unannotated reference type is initialized with, or assigned a value of, `null`.
+- A warning shall be generated when a variable of a nullable or non-nullable reference type, whose value might be `null`, is dereferenced; otherwise, that variable may be dereferenced safely.
+- The null-forgiving operator `!` suppresses warnings about dereferencing a possible null reference.
+
+> *Note*: A Nullable context (§Nullable-Contexts) determines if reference types are considered nullable and non-nullable, or null-oblivious. Switching from a disabled nullable context to an enabled nullable context has the effect of changing the type of reference variables from *null-oblivious* to *non-nullable*. *end note*
+<!-- markdownlint-disable MD028 -->
+
+<!-- markdownlint-enable MD028 -->
+> *Example*:
+>
+> <!-- Example: {template:"code-in-main-without-using", name:"NullableAnnotationContext2", ignoredWarnings:["CS0219"], expectedWarnings:["CS8600","CS8600","CS8602"], expectedException:"NullReferenceException"} -->
+> ```csharp
+> #nullable enable
+> string? s1 = null;    // OK; ? makes s2 nullable
+> string s2 = null;     // Warning; s1 is non-nullable
+> s2 = null;            // Warning; s1 is non-nullable
+> char c1 = s2[1];      // Warning; dereference of a possible null; throws NullReferenceException
+> c1 = s2![1];          // The warning is suppressed
+> ```
+>
+> *end example*
+
+#### §Nullable-Warning-Context Nullable warning context
+
+This context controls whether nullable warnings are issued by the compiler.
+
+The compiler uses static flow analysis to determine the null state of any reference variable. When the nullable warning context is enabled, a reference variable’s null state (§Nullabilities-And-Null-States) is either *not null* or *maybe null*, and
```

Comment:
> The compiler uses static flow analysis […]

Is that a requirement of the Standard? I’m a bit fuzzy on where the (fuzzy) line is here.

As I currently understand a compiler must accept the annotations and pragmas, but may produce any warnings based on those that it likes – specific warnings are not being required.

The hope/allowance is that over time compilers will produce better warnings (and no “hallucinations” to use the in-phrase ;-)) from improving their analyses. But this also implies a compiler could choose to produce no warnings and still be compliant, and if it chooses to do so no analyses would be required…

That the analysis must be *static* could be questioned (consider typing where some limited dynamic typing is supported) but I expect it's unlikely to be an issue for the foreseeable future. But what about it being a *flow* analysis – could a new algorithm be developed that doesn’t fit that category?

TL;DR: I wondering whether, given the lack of requirements to meet conformance for this feature, why the language is not about “a compiler *may*” do such and such, like use a static flow analysis… produce a warning… or even determine that something definitely *is-null* and not just *maybe-null*